R/end_rains.R

```diff
@@ -24,85 +24,76 @@ end_rains <- function(data, date_time, station = NULL, year = NULL, rain = NULL,
-    dplyr::filter((roll_sum_rain > min_rainfall) | is.na(x=roll_sum_rain)) %>% 
-    dplyr::group_by(.data[[year]], .add = TRUE, .drop = drop) %>%
-    dplyr::filter(.data[[doy]] >= start_day & .data[[doy]] <= end_day, .preserve = TRUE)
-  
-  if (output == "doy"){
-    end_of_rains <- end_of_rains %>%
-      dplyr::summarise(end_rains = ifelse(is.na(x = dplyr::last(x = roll_sum_rain)), 
-                                         NA, 
-                                         dplyr::last(x=.data[[doy]], default=NA)))
     
-  } else if (output == "date") {
-    end_of_rains <- end_of_rains %>%
-      dplyr::summarise(end_rains = dplyr::if_else(is.na(x = dplyr::last(x = roll_sum_rain)), 
-                                                 as.Date(NA),
-                                                 dplyr::last(.data[[date_time]], default=NA)))
-  } else {
-    end_of_rains <- end_of_rains %>%
-      dplyr::summarise(end_rains_doy = ifelse(is.na(x = dplyr::last(x = roll_sum_rain)), 
-                                             NA, 
-                                             dplyr::last(x=.data[[doy]], default=NA)),
-                       end_rains_date = dplyr::if_else(is.na(x = dplyr::last(x = roll_sum_rain)), 
-                                                      as.Date(NA),
-                                                      dplyr::last(.data[[date_time]], default=NA)))
-  }
-  return(end_of_rains)
-}
+    # 4. From here we are using the End of Rains Dialog R Code, so this can all be replaced with that code directly
+    # to avoid dropping levels, set as factor
+    year_type <- data_book$get_column_data_types(data_name=data, columns=year)
+    data_book$convert_column_to_type(data_name=data, col_names=year, to_type="factor")
+    data_book$convert_linked_variable(from_data_frame=data, link_cols=c(year))
+    roll_sum_rain <- instatCalculations::instat_calculation$new(type="calculation", function_exp="RcppRoll::roll_sumr(x=rain, n = interval_length, fill=NA, na.rm=FALSE)", result_name="roll_sum_rain", calculated_from=setNames(list(rain), data))
+    conditions_filter <- instatCalculations::instat_calculation$new(type="filter", function_exp="(roll_sum_rain > min_rainfall) | is.na(x=roll_sum_rain)", sub_calculations=list(roll_sum_rain))
+    grouping_by_station_year <- instatCalculations::instat_calculation$new(type="by", calculated_from=setNames(list(year), data))
+    doy_filter <- instatCalculations::instat_calculation$new(type="filter", function_exp="doy >= start_day & doy <= end_day", calculated_from=databook::calc_from_convert(x=list(data=doy)))
+    
+    # If "DOY (End Rains) is checked then run this line:
+    if (output == "day"){
+      end_rains <- instatCalculations::instat_calculation$new(type="summary", function_exp="ifelse(test=is.na(x=dplyr::last(x=roll_sum_rain)), yes=NA, no=dplyr::last(x=doy))", result_name="end_rains", calculated_from=setNames(list(doy), data), save=2)
```

Comment:
Oh nice one that you worked out that we need to do `setNames` here. Do you know why we have to do this here? Happy to explain, but perhaps you already know! 

In R-Instat we do

`calculated_from = list(data = "doy")`

But since `data` is a parameter, we have to do `setNames` so that it runs it correctly. Otherwise, it would just use the literal term `data`, and not clock it is a parameter we want to read in.

Not sure why I'm telling you this since you've worked it out. I just realised when I saw your message in this issue that I forgot to mention that.

Reply:
Yep this line is all good!